utils/src/main/java/software/amazon/awssdk/utils/async/ByteBufferStoringSubscriber.java

```diff
@@ -0,0 +1,194 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+package software.amazon.awssdk.utils.async;
+
+import static software.amazon.awssdk.utils.async.StoringSubscriber.EventType.VALUE;
+
+import java.nio.ByteBuffer;
+import java.util.concurrent.atomic.AtomicInteger;
+import org.reactivestreams.Subscriber;
+import org.reactivestreams.Subscription;
+import software.amazon.awssdk.annotations.SdkProtectedApi;
+import software.amazon.awssdk.utils.Validate;
+import software.amazon.awssdk.utils.async.StoringSubscriber.Event;
+
+/**
+ * An implementation of {@link Subscriber} that stores {@link ByteBuffer} events it receives for retrieval.
+ *
+ * <p>Stored bytes can be read via {@link #transferTo(ByteBuffer)}.
+ */
+@SdkProtectedApi
+public class ByteBufferStoringSubscriber implements Subscriber<ByteBuffer> {
+    /**
+     * The minimum amount of data (in bytes) that should be buffered in memory at a time. The subscriber will request new byte
+     * buffers from upstream until the bytes received equals or exceeds this value.
+     */
+    private final int minimumBytesBuffered;
```

Comment:
Should it be long?

Reply:
I was thinking 4gb should be plenty, but maybe you're right.